### ISSUE TITLE
Consider renaming "to" in "safeTransfer" flows to remind about unsafe external call

### ISSUE THREAD

**Issue by ethosdev** [OpenZeppelin/openzeppelin-contracts#2822]
An unsafe external call is made to innocent looking `to`.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/b174f067df18e3214f290c62d50f6721e87c05cf/contracts/token/ERC721/ERC721.sol#L371

https://www.paradigm.xyz/2021/08/the-dangers-of-surprising-code/

A suggestion that comes to mind is to use "untrustedTo" instead of "to", but any other name would be a good reminder about the dangers of untrusted external calls.

The rename should propagate back to the safeTransfer functions like https://github.com/OpenZeppelin/openzeppelin-contracts/blob/b174f067df18e3214f290c62d50f6721e87c05cf/contracts/token/ERC721/ERC721.sol#L168

**Comment by frangio** [comment]
Thanks for the suggestion @ethosdev. We're interested in ways to make this risk clearer. However, I'm not sure that this rename accomplishes that. In fact it confused me and made me question whether `to` is actually untrusted or rather trusted, and I think it's neither/both.

Renaming this to `untrustedTo` could imply that it is OK to pass in an untrusted address, when in fact it's the opposite: an untrusted address introduces risk of reentrancy. `trustedTo` is also not accurate because it is perfectly fine to pass in an untrusted address, as long as you invoke the function such that reentrancy poses no risk (e.g. following checks-effects-interactions).

Whether `to` is trusted or untrusted, and marking it as such, should be done at the caller site it seems, as a hint to the reader that there is or isn't risk of reentrancy (notwithstanding the unfortunate "safe" terminology).

**Comment by QEDK** [comment]
> Thanks for the suggestion @ethosdev. We're interested in ways to make this risk clearer. However, I'm not sure that this rename accomplishes that. In fact it confused me and made me question whether `to` is actually untrusted or rather trusted, and I think it's neither/both.
> 
> Renaming this to `untrustedTo` could imply that it is OK to pass in an untrusted address, when in fact it's the opposite: an untrusted address introduces risk of reentrancy. `trustedTo` is also not accurate because it is perfectly fine to pass in an untrusted address, as long as you invoke the function such that reentrancy poses no risk (e.g. following checks-effects-interactions).
> 
> Whether `to` is trusted or untrusted, and marking it as such, should be done at the caller site it seems, as a hint to the reader that there is or isn't risk of reentrancy (notwithstanding the unfortunate "safe" terminology).

Agreed, since in reality, `safeTransfer` is only _safer_ at best. I don't think renaming will help, it's best to cover or link possible dangers in documentation and make the developer community aware of what exactly future risks can be.

**Comment by ethosdev** [comment]
Thanks for the responses @frangio and @QEDK 

> Renaming this to `untrustedTo` could imply that it is OK to pass in an untrusted address, when in fact it's the opposite: an untrusted address introduces risk of reentrancy. 

Indeed the word trusted is still not clear enough.  What do you think about `unsafeTo`?

While the `to` could be safe in some cases, the library errs to bring most alert to devs and readers to be careful.

We thought we'd all be better at watching out for unsafe external calls, but even 5 years after TheDAO, they are very good at hiding behind variable names.

I understand the objections that other variables names are not as elegant as `to`, but whenever we do things indirectly, humans can always forget.  For example, in one context you tell yourself that I know a particular `to` is safe, and later on you see another `to` and also incorrectly assume it is safe.

I also know we cannot solve for everything.  This suggestion is to consider if we can do more about `to` instead of just putting it as a code comment.

> should be done at the caller site

I also agrees with this.  Much less control over this though, and I think it's a good opportunity for OpenZeppelin to set a good example and reminders about unsafe external calls and how easily they hide behind variable names.

**Comment by ethosdev** [comment]
> `safeTransfer` is only _safer_ at best.

I spoke to someone who thought it's better not to call these methods "safe". Someone could open a PR against the ERC and for now I am not ready for those discussions.

> I don't think renaming will help

Is this really the concern?  (If an external call is being made to a variable called something like `unsafeTo` or `unsafeRecipient`, doesn't that bring much more attention and alert than if it was called with the shorter name?)

> make the developer community aware of what exactly future risks can be

The lesson behind TheDAO was not specifically about reentrancy.  It was about external calls and how they can be unsafe, and it's #1 danger in smart contracts:

https://consensys.github.io/smart-contract-best-practices/recommendations/#external-calls

It's been 5 years and I think we can improve here on the actual variable names, not just adding more comments and documentation that people can more easily forget about.
